DDD-9.md

```diff
@@ -0,0 +1,111 @@
+# DDD-9: A buffer-less implementation of Oracle LogMiner
+
+## Motivation
+
+While most Debezium Oracle users rely on the heap-based implementation where the transaction data is cached in the JVM's heap, this puts a substantially higher burden of resources on the Kafka Connect cluster.
+In addition, users that maintain the connector and Kafka Connect environments may not have any knowledge of what the transaction workload is like, which makes scaling the JVM heap difficult.
+
+For users who do use the off-heap implementations of Infinispan and Ehcache, these often come at the cost of performance.
+Each of these off-heap caches implement some level of disk access, which will naturally be orders of magnitude slower than reading/writing directly to JVM heap.
+
+Lastly, one of the more technical critiques is everyone believes the Oracle connector's code is significantly complex.
+If we eliminate the need to buffer transactions entirely, that significantly reduces the complexity and integration with cache providers, simplifies the configuration, and slims the documentation so that perhaps it's easier to digest with less moving parts. 
+
+## Goals
+
+* Reduce code complexity, eliminating the need for on-heap and off-heap cache integrations.
+* Reduce connector memory footprint
+* Avoid issues with split transactions that yield _unsupported_ operations.
+* Provides an easy way to manage transaction-level sequences for restarts.
+* Remove the need for `rac.nodes`.
+* Remove the need for a flush table and flush strategies.
+
+## Additional benefits
+
+* Data set would be provided in committed, chronological order.
+* Implementation can use simple read, parse, dispatch pattern similar to other connectors.
+* Read strategy aligns closely with the current one used when `lob.enabled` is set to `true`.
+
+## Requirements
+
+* LogMiner session must be started to return only committed data.
+* The LogMiner resume SCN is used to control the low watermark in the offsets.
+* Transactions committed after the resume SCN must have its id and SCN maintained in the offsets.
+* LogMiner query should exclude any transaction already committed since the resume SCN.
+
+## Proposed changes
+
+### Offsets
+
+Currently, the Oracle connector maintains two critical values in the offsets, `scn` and `commit_scn`, shown here:
+
+```json
+{
+    "scn": "123456",
+    "commit_scn": "123456:1:txId1-txId2,123457:2:txId4"
+}
+```
+
+Moving forward, the connector needs to maintain the resume SCN and each committed transaction id since the resume SCN and its commit SCN.
+In addition, need to maintain the last processed transaction id and event sequence.
+I would also prefer to move away from _encoded_ values to the more natural nested `Struct`s style.
+
+I'm not sure how hard this makes editing the offsets for users, but this what I have in mind:
```

Comment:
Offset is just a `Map`. This means you'd need to generate the JSON/structure yourself and put it under a single key.
I don't think that map of maps is supported.

Reply:
In that case, whether we use an encoded string or JSON is just a mere implementation detail and boils down to what is easier for users to supply in the end.